stan/math/prim/fun/multiply.hpp

```diff
@@ -23,8 +23,9 @@ namespace math {
  * @param c scalar
  * @return product of matrix and scalar
  */
-template <typename Mat, typename Scal, require_eigen_t<Mat>* = nullptr,
-          require_stan_scalar_t<Scal>* = nullptr>
+template <typename Mat, typename Scal, require_stan_scalar_t<Scal>* = nullptr,
+          require_eigen_t<Mat>* = nullptr,
+          require_not_return_type_t<is_var, Scal, Mat>* = nullptr>
```

Comment:
mostly I'm just not a huge fan of `require_all_not_st_*`. The logic is sort of weird to me. Using the return type feels nice to me because it's most directly saying, "If the return type is X use this version"

Reply:
I on the other hand find the return type logic a bit confusing. What overload is used depends on arguments, not return value.   Also the only think I know from the top of my head about return type metaprogram is that it is confusing and I have to look up what does it do every time I use it or read code that uses it.

Reply:
I prefer the `require_all_not_st_*` logic as well and so I'll probably just use that.

Reply:
Huh! Yeah I mean if you both prefer `require_all_not_st_*` then I'm fine with changing these. I always think of `return_type_t<T1, T2, ...>` as "run `scalar_type_t<>` over each input then select the least upper bound of all the types". So all containers go down to their scalar types, and then the return type is double if all double, var if any are var, `fvar<double>` if any are `fvar<double>`. The only weirdo is stuff like `double, var, fvar<double>` which I'm pretty sure returns `fvar<var>`

Reply:
Well crap, is there any reason to keep this if `require_all_not_st_*` already does the job?

I am happy to keep it if we need it at some vague point in the future, but I guess now is the time to cut it if we do not.

Reply:
Uhh, I have tests for it so let's keep it for now